// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
edition = "2023";

package koladata.ext.persisted_data;

import "google/protobuf/timestamp.proto";

message PersistedIncrementalDataSliceManagerMetadata {
  // The version of the manager. Always set. Example format: "1.0.0"
  string version = 1;

  // The type of the initial data manager.
  string initial_data_manager_id = 4;

  // The number of times the metadata was updated. Always set. Starts at 0 and
  // is incremented by 1 with each update.
  // This is used to make sure that updates to the manager are transactional.
  // It helps to detect concurrent writers and prevents them from overwriting
  // each other's changes.
  int64 metadata_update_number = 3;

  // Metadata about all the revisions. The order here reflects the chronological
  // order in which the revisions were created. The revisions correspond to
  // operations with side-effects on the schema and/or the data managed by the
  // manager.
  repeated RevisionMetadata revision_history = 2;
}

message RevisionMetadata {
  // The UTC timestamp of the revision.
  google.protobuf.Timestamp timestamp = 1;

  // A description of the revision. Optional. Typically provided by the user,
  // although some operations may have descriptions that are programmatically
  // generated.
  string description = 2;

  // The names of the data bags that were added to the various bag managers by
  // this revision. These fields should be understood as sets; the order of the
  // elements does not have any significance. It is the responsibility of each
  // bag manager to remember the order in which the bags were added.
  repeated string added_data_bag_names = 3;
  repeated string added_schema_bag_names = 4;
  repeated string added_snn_to_data_bags_update_bag_names = 5;

  // More information about the specific kind of revision.
  oneof revision_kind {
    CreationMetadata creation = 6;
    AttributeUpdateMetadata attribute_update = 7;
    BranchMetadata branch = 8;
  }
}

message CreationMetadata {}

message AttributeUpdateMetadata {
  // The data slice path at which the attribute was updated. Produced by
  // DataSlicePath.to_string(). Example: ".query[:]"
  string at_path = 1;

  // The name of the attribute that was updated.
  // Example: for at_path = ".query[:]", attr_name = "query_text" or attr_name =
  // "doc". Since the attribute value can contain entities with attributes of
  // their own, the update can affect more attributes than just attr_name, which
  // simply records the attr_name argument of the update call.
  string attr_name = 2;
}

message BranchMetadata {
  // The persistence directory from which the branch was created. It is an
  // absolute path.
  string parent_persistence_directory = 1;

  // The index of the revision in the parent manager's revision_history on top
  // of which the branch was created. Always >= 0, so the index is always
  // absolute and never relative to the last revision. Relative indices would be
  // ambiguous because one can create more revisions of the parent manager after
  // the branch is created.
  int32 parent_revision_history_index = 2;
}
